include/vapor/DCUGRID.h

```diff
@@ -0,0 +1,84 @@
+#include <vector>
+#include <algorithm>
+#include <map>
+#include <iostream>
+#include <vapor/MyBase.h>
+#include <vapor/NetCDFCFCollection.h>
+#include <vapor/DerivedVar.h>
+#include <vapor/DerivedVarMgr.h>
+#include <vapor/UDUnitsClass.h>
+#include <vapor/utils.h>
+#include <vapor/DCCF.h>
+
+#pragma once
+
+namespace VAPoR {
+
+//!
+//! \class DCUGRID
+//! \ingroup Public_VDCUGRID
+//!
+//! \brief Class for reading a UGRID data set
+//! stored as a series
+//! of NetCDF files: https://ugrid-conventions.github.io/ugrid-conventions/
+//!
+//! \author John Clyne
+//! \date    July, 2021
+//!
+class VDF_API DCUGRID : public VAPoR::DCCF {
+public:
+    int OpenVariableRead(size_t ts, string varname, int level = 0, int lod = 0) override;
+
+    int Read(int fd, int *data) override;
+    int ReadRegion(int fd, const vector<size_t> &min, const vector<size_t> &max, int *data) override;
+
+    int CloseVariable(int fd) override;
+
+protected:
+    int initialize(const vector<string> &paths, const std::vector<string> &options) override;
+
+    int initAuxilliaryVars(NetCDFCFCollection *ncdfc, std::map<string, DC::AuxVar> &auxVarsMap) override;
+
+    int initDataVars(NetCDFCFCollection *ncdfc, std::map<string, DC::DataVar> &dataVarsMap) override;
+
+    int initMesh(NetCDFCFCollection *ncdfc, std::map<string, DC::Mesh> &meshMap) override;
+
+    string getMapProjection() const override { return ("+proj=eqc +ellps=WGS84 +lon_0=0.0 +lat_0=0.0"); }
```

Comment:
How about we put the function implementation in the `.cpp` file?

Otherwise, making a change to the function body in the header file will cause a re-compilation of almost the entire project.

Reply:
Putting inexpensive getters in .h files is done throughout the code base, and widely done in practice. What else would be gained by moving the definition other than saving a few microseconds (at most) during compilation?

Reply:
I'm afraid the save is more than microseconds. 
-- With implementation being in the header, every file that depends on DC will be re-compiled. That's almost the entire VAPOR, and that compilation is in the order of minutes. 
-- With implementation being in the body, only DC is re-compiled and everybody else just need to re-link, which takes microseconds. 

Plus, it really makes the header look tidy and easier to read when you hide details away.

Reply:
> I'm afraid the save is more than microseconds.
> -- With implementation being in the header, every file that includes this header will be re-compiled. That's almost the entire VAPOR, and that compilation is in the order of minutes.
> -- With implementation being in the body, only DC is re-compiled and everybody else just need to re-link, which takes microseconds.
> 

I don't believe that the compilation time will change significantly if the definition is moved. You'll have to do an experiment and convince me with the results :-)

> Plus, it really makes the header look tidy and easier to read when you hide details away.

That's your personal opinion. 

@shaomeng I am not just being stubborn.  I'm making a point. Changes requested by reviewers should have a meaningful impact to the code in terms of performance, readability, robustness, future-proofing, following best practices, or simply following our own conventions. I don't want developers to have to change their code simply because a reviewer has an opinion on how things should look that may or may not agree with the opinion of others. That just waists everyone's time and does not lead to better code.

Reply:
Regarding compilation time, I'm *not* saying that the 1st time you compile the project there will be a significant difference. I'm saying that when somebody debugs the code, for example change the function implementation or simply insert a print statement, function body being in the header will trigger all the downstream files to be re-compiled again. That re-compile of all the downstream files cost minutes, and that cost is totally avoidable. 

I can tell my personal experience debugging a crashing problem in the last release cycle. I needed to insert a few print statements in some grid classes. Guess how much time it takes to re-compile a few print statements? Minutes. Why? Almost the entire project is re-compiled, because almost every file includes a grid class.

Reply:
Regarding where to put function implementations, I don't think it's my personal opinion. It's well established C++ practice, with *real* benefits (e.g. the case I raised above). Moreover, it's stated in VAPOR coding convention.

Reply:
Regarding opinion or not and in general how to review code and more general how to produce better code, we'll probably need a deeper discussion. There is definitely great discrepancy between one review and another review. For example, I was asked to not use trailing returns in a PR because it was thought by some developers that trailing returns are difficult to read.

Reply:
> Regarding where to put function implementations, I don't think it's my personal opinion. It's well established C++ practice, with _real_ benefits (e.g. the case I raised above). Moreover, it's stated in VAPOR coding convention.

I should have been more clear about what is reasonable to put into a header file. I was not suggesting that complicated code which might require debugging (print statements) is reasonable to put in a .h file. But simple statements such as those that define many getters functions should not require debugging. The case in point returns a string literal. Something is really going off the rails if you have to put a print statement inside a single line function that returns a constant value :-). Moreover, I'm not aware of any well established C++ practice that recommends against simple expressions in .h files. Do you have a reference for this? Qt, Vtk, and Google Open Source all have lots of .h files with one-line functions (and more) defined in them. The Google Source Chromium guide simply says "Avoid this [definitions] in header files except for cheap non-virtual getters and setters. ", which seems reasonable to me.

Reply:
> Regarding where to put function implementations, I don't think it's my personal opinion. It's well established C++ practice, with _real_ benefits (e.g. the case I raised above). Moreover, it's stated in VAPOR coding convention.

I don't see anything in our coding [conventions](https://github.com/NCAR/VAPOR/wiki/Vapor-Coding-Convention) about this issue.

Reply:
> ause it was thought by some developers that trailing returns are difficult to read. Is that an opinion or not?

I remember this discussion and I think at the time we were not aware that trailing returns actually had some practical benefit (not just syntactical sugar): they allow return types to be auto deduced by the compiler. I'd certainly be willing to revisit the issue.

Reply:
Well, the separation of .h and .cpp files is something taught at every college class, for example, [this one](http://websites.umich.edu/~eecs381/handouts/CppHeaderFileGuidelines.pdf). I can only imagine that some guidelines don't specify it in explicit terms is because it's well accepted that there's no need to specify it.

Coming back to that Google document, it specifies non-virtual functions. But the function in question here has `override` in its signature, indicating a virtual function.

Lastly, I think it's a bad assumption to make that "I don't need to come back and debug this function." As a matter of fact, I personally needed to change function bodies living in headers for debugging purposes in the last release cycle.

Reply:
> I don't see anything in our coding [conventions](https://github.com/NCAR/VAPOR/wiki/Vapor-Coding-Convention) about this issue.

I don't think this sentence means anything else given its context.
`Any public class should generally be declared in its own header file, and defined in its own source file.` 
There's no mention "if you meet condition 1, 2, and 3, then you can leave the definition in the header.

Reply:
> I remember this discussion and I think at the time we were not aware that trailing returns actually had some practical benefit (not just syntactical sugar): they allow return types to be auto deduced by the compiler. I'd certainly be willing to revisit the issue.

I didn't try to raise attention to the benefit (or lack of benefit) of trailing returns (though I welcome a discussion on it). I was trying to raise the point of consistency when exercising code review activities. I'd agree that personal opinion isn't a valid basis for change requests. I also think consistently applying a certain standard is important.

Reply:
@shaomeng, I'm sorry but you haven't convinced me that there is a good reason to prohibit simple member functions from being defined in a .h file. There appears to be no practical reason for such a prohibition. I don't really believe that there is a need for debugging a function that does no more than return a literal of even a member variable., do you? You've mentioned having trouble debugging more complex functions defined in .h files. If so, I'd suggest submitting an issue, or even a PR, to move those troublesome functions. There also does not appear to be any consensus amongst the C++ community that I can see. You've claimed that every college class teaches not to do this, but that claim seems wholly unsubstantiated. The existence of a single example doesn't do much to support this argument. On the contrary I've identified three large, modern C++ development efforts that do not abide by any such convention (add a fourth, Intel OSPRay to the list as well). And, FWIW the Google C++ Style Guide recommends limiting functions defined in .h files to 10 lines of code. 

If you really think the VAPOR code base would be improved by adopting a convention on this that we all abide by, you should propose it to the group as a change to our coding conventions. If you can convince the majority I am happy to comply (and will expect others to as well).

Reply:
I'm sorry John, I think the VAPOR coding convention has a clear specification, but my interpretation could be wrong. What do you think this sentence in it means? 
`Any public class should generally be declared in its own header file, and defined in its own source file.`
If it doesn't mean every function body goes to .cpp files, I'm happy to learn what the guideline actually is regarding what to put in headers and in .cpp files. 

(Note, what this specification means, and if real code abides to it, are two different issues.)

Reply:
Sam, there is a qualifier in the statement that may be leading to the confusion: "generally". In a context like this "generally" means that most of the time the rule applies, but "generally" provides the flexility to ignore it. Lacking specificity about when the rule can be ignored it is left up to the judgement of the reader.

Reply:
Well, talking about "generally." When judging how and when "generally" applies, my understanding is that there should be some strong and evident benefits to support an occasional break of the rule. That strong and evident benefit should be in the nature of enhancing the functionality, robustness, portability (and alike) of the program. One example of such scenario is templated function.

The justification to apply "generally," however, should *not* be: "I don't want to spend 5 extra seconds to abide to the rules." Because, that's almost the definition of abusing the flexibility provided by the rules. Do you think this is a reasonable understanding of "generality" in that sentence, and how to interpret rules in general?

Regard to this specific case, I haven't seen any benefit to the software that you provide to support this occasion of breaking the rule. If I missed it, maybe you could raise it again?

Reply:
@shaomeng I'm not debating this anymore. Since there does not appear to be any consensus from the C++ community if you want to ban all method definitions from .h files you'll need to convince the group to clearly state this in our coding guidelines.